### PR TITLE
CORDA-530 Unduplicate code

### PR DESCRIPTION
**Pull request by andr3ej**

### CODE REVIEW

**Comment by josecoll** [comment]
What do any of these changes have to do with **Vault soft locking optimisations**???

**Comment by mikehearn** [comment]
Seems like cleanups done as Andrzej looked at the code. I haven't done a thorough review, but most of the changes look reasonable: either small style changes or pulling db transactions up into larger scopes. Is that right? A better description of the changes and what motivated them would be useful here @andr3ej

**Comment by andr3ej** [comment]
Summary of changes:
* duplicated code no longer duplicated
* session no longer acquired via static state where it wasn't necessary to do so

Motivation:
* refactoring while familiarising myself with soft locking - doing so keeps tech debt under control, while restricting it to the area i'm working on avoids going down rabbit holes that may have to be reverted later, also see http://www.extremeprogramming.org/rules/refactor.html which would be good to sticky somewhere

**Comment by andr3ej** [comment]
@josecoll @mikehearn anything else needed on this PR?

**Comment by andr3ej** [comment]
> andr3ej commented 5 days ago 
> @josecoll @mikehearn anything else needed on this PR?

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt [hunk 1]
```diff
@@ -270,24 +263,22 @@ open class PersistentNetworkMapCache(private val serviceHub: ServiceHubInternal)
     /**
      * Load NetworkMap data from the database if present. Node can start without having NetworkMapService configured.
      */
-    private fun loadFromDB() {
+    private fun loadFromDB(session: Session) {
```
Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: Why are you passing a `Session` around now?

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt [hunk 2]
```diff
@@ -311,11 +302,9 @@ open class PersistentNetworkMapCache(private val serviceHub: ServiceHubInternal)
...
-        createSession {
-            val info = findByIdentityKey(it, nodeInfo.legalIdentitiesAndCerts.first().owningKey).single()
-            it.remove(info)
-        }
+    private fun removeInfoDB(session: Session, nodeInfo: NodeInfo) {
+        val info = findByIdentityKey(session, nodeInfo.legalIdentitiesAndCerts.first().owningKey).single()
+        session.remove(info)
     }
 
     private fun findByIdentityKey(session: Session, identityKey: PublicKey): List<NodeInfoSchemaV1.PersistentNodeInfo> {
```
Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: Seems you are passing a `Session` around to all private methods now.

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt [hunk 3]
```diff
@@ -326,48 +315,40 @@ open class PersistentNetworkMapCache(private val serviceHub: ServiceHubInternal)
         return query.resultList
     }
 
-    private fun queryByIdentityKey(identityKey: PublicKey): List<NodeInfo> {
-        createSession {
-            val result = findByIdentityKey(it, identityKey)
-            return result.map { it.toNodeInfo() }
-        }
+    private fun queryByIdentityKey(session: Session, identityKey: PublicKey): List<NodeInfo> {
```
Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: ditto

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt [hunk 4]
```diff
@@ -326,48 +315,40 @@ open class PersistentNetworkMapCache(private val serviceHub: ServiceHubInternal)
...
-            val query = it.createQuery(
-                    // We do the JOIN here to restrict results to those present in the network map
-                    "SELECT DISTINCT l FROM ${NodeInfoSchemaV1.PersistentNodeInfo::class.java.name} n JOIN n.legalIdentitiesAndCerts l WHERE l.name = :name",
-                    NodeInfoSchemaV1.DBPartyAndCertificate::class.java)
-            query.setParameter("name", name.toString())
-            val candidates = query.resultList.map { it.toLegalIdentityAndCert() }
-            // The map is restricted to holding a single identity for any X.500 name, so firstOrNull() is correct here.
-            return candidates.firstOrNull()
-        }
+    private fun queryIdentityByLegalName(session: Session, name: CordaX500Name): PartyAndCertificate? {
```
Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: ditto

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt [hunk 5]
```diff
@@ -326,48 +315,40 @@ open class PersistentNetworkMapCache(private val serviceHub: ServiceHubInternal)
...
-    private fun queryByLegalName(name: CordaX500Name): List<NodeInfo> {
-        createSession {
-            val query = it.createQuery(
-                    "SELECT n FROM ${NodeInfoSchemaV1.PersistentNodeInfo::class.java.name} n JOIN n.legalIdentitiesAndCerts l WHERE l.name = :name",
-                    NodeInfoSchemaV1.PersistentNodeInfo::class.java)
-            query.setParameter("name", name.toString())
-            val result = query.resultList
-            return result.map { it.toNodeInfo() }
-        }
+    private fun queryByLegalName(session: Session, name: CordaX500Name): List<NodeInfo> {
```
Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: ditto

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt [hunk 6]
```diff
@@ -326,48 +315,40 @@ open class PersistentNetworkMapCache(private val serviceHub: ServiceHubInternal)
...
-            val query = it.createQuery(
-                    "SELECT n FROM ${NodeInfoSchemaV1.PersistentNodeInfo::class.java.name} n JOIN n.addresses a WHERE a.pk.host = :host AND a.pk.port = :port",
-                    NodeInfoSchemaV1.PersistentNodeInfo::class.java)
-            query.setParameter("host", hostAndPort.host)
-            query.setParameter("port", hostAndPort.port)
-            val result = query.resultList
-            return if (result.isEmpty()) null
-            else result.map { it.toNodeInfo() }.singleOrNull() ?: throw IllegalStateException("More than one node with the same host and port")
-        }
+    private fun queryByAddress(session: Session, hostAndPort: NetworkHostAndPort): NodeInfo? {
```
Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: ditto

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt (hunk 1 above)
Comment (andr3ej) [reply node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: There's no need to obtain it via the static threadlocal interface if we already have it.

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt (hunk 3 above)
Comment (andr3ej) [reply node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: These dittos add no value to the review, I've seen the diff.

Comment (josecoll) [reply node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: The dittos highlight you are passing a variable around 7 times.

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt (hunk 1 above)
Comment (josecoll) [reply node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: So why not use a private global variable?

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt (hunk 3 above)
Comment (andr3ej) [reply node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: * then say 7 times
* the param replaces createSession, which doesn't even create a session

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt (hunk 1 above)
Comment (andr3ej) [reply node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: the session ceases to be valid at the end of the transaction so it absolutely does not belong in a global variable

Comment (Clintonio) [reply node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: I think it makes sense to make your parameters explicit if they are not member variables. If they're functions that update the DB and need a session it makes sense that they take it as a parameter.

File: node/src/main/kotlin/net/corda/node/utilities/DatabaseTransactionManager.kt [hunk 7]
```diff
@@ -62,6 +62,7 @@ class DatabaseTransaction(isolation: Int, val threadLocal: ThreadLocal<DatabaseT
     }
 }
 
+fun currentSession() = DatabaseTransactionManager.current().session
```
Comment (Clintonio) [inline node/src/main/kotlin/net/corda/node/utilities/DatabaseTransactionManager.kt]: Rename to currentDBSession perhaps? It's not obvious from first glance what kind of session this is.

Comment (andr3ej) [reply node/src/main/kotlin/net/corda/node/utilities/DatabaseTransactionManager.kt]: Done

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt (hunk 1 above)
Comment (josecoll) [reply node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: Wow - thats powerful stuff. I didn't realise you could directly access the delegated property from the enclosing transaction (lambda).
eg.    **session** in the following code snippet:
`init {
        loadFromFiles()
        serviceHub.database.transaction { loadFromDB(session) }
    }`

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt [hunk 8]
```diff
@@ -31,7 +31,6 @@ import net.corda.node.services.messaging.sendRequest
 import net.corda.node.services.network.NetworkMapService.FetchMapResponse
 import net.corda.node.services.network.NetworkMapService.SubscribeResponse
 import net.corda.node.utilities.AddOrRemove
-import net.corda.node.utilities.DatabaseTransactionManager
 import net.corda.node.utilities.bufferUntilDatabaseCommit
 import net.corda.node.utilities.wrapWithDatabaseTransaction
 import org.hibernate.Session
```
Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: Unused import on line 41 (which bizarrely I cannot annotate)
`import java.time.Duration`

File: node/src/test/kotlin/net/corda/node/services/persistence/NodeAttachmentStorageTest.kt [hunk 9]
```diff
@@ -99,18 +98,18 @@ class NodeAttachmentStorageTest {
...
-
-                    // Corrupt the file in the store.
-                    val bytes = testJar.readAll()
-                    val corruptBytes = "arggghhhh".toByteArray()
-                    System.arraycopy(corruptBytes, 0, bytes, 0, corruptBytes.size)
-                    val corruptAttachment = NodeAttachmentService.DBAttachment(attId = id.toString(), content = bytes)
-                    DatabaseTransactionManager.current().session.merge(corruptAttachment)
-                    id
-                }
+        database.transaction {
```
Comment (josecoll) [inline node/src/test/kotlin/net/corda/node/services/persistence/NodeAttachmentStorageTest.kt]: Original indentation was correct.

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt (hunk 8 above)
Comment (andr3ej) [reply node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: Done

File: node/src/test/kotlin/net/corda/node/services/persistence/NodeAttachmentStorageTest.kt (hunk 9 above)
Comment (andr3ej) [reply node/src/test/kotlin/net/corda/node/services/persistence/NodeAttachmentStorageTest.kt]: This was probably affected by the recent mass reformat. I've now moved database.transaction onto the same line as val id to make it less weird.

**Pull request merged**